private-identity-audit/src/proofs.rs

```diff
@@ -17,26 +17,150 @@
 
 use crate::errors::{ErrorKind, Fallible};
 use crate::Challenge;
-use cryptography_core::curve25519_dalek::{ristretto::RistrettoPoint, scalar::Scalar};
+use cryptography_core::curve25519_dalek::{
+    ristretto::{CompressedRistretto, RistrettoPoint},
+    scalar::Scalar,
+};
 use rand_core::{CryptoRng, RngCore};
 
-#[derive(Clone)]
+use codec::{Decode, Encode, Error as CodecError, Input, Output};
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+use sp_std::vec::Vec;
+
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub struct InitialMessage {
     a: RistrettoPoint,
     pub generators: Vec<RistrettoPoint>,
 }
 
-#[derive(Clone)]
+impl Encode for InitialMessage {
+    #[inline]
+    fn size_hint(&self) -> usize {
+        32 + self.generators.len() * 32
+    }
+
+    fn encode_to<W: Output>(&self, dest: &mut W) {
+        let a = self.a.compress();
+        let generators: Vec<[u8; 32]> = self
+            .generators
+            .clone()
+            .into_iter()
+            .map(|g| *g.compress().as_bytes())
+            .collect::<Vec<_>>();
+
+        a.as_bytes().encode_to(dest);
+        generators.encode_to(dest);
+    }
+}
+
+impl Decode for InitialMessage {
+    fn decode<I: Input>(input: &mut I) -> Result<Self, CodecError> {
+        let (a, generators) = <([u8; 32], Vec<[u8; 32]>)>::decode(input)?;
+        let a = CompressedRistretto(a)
+            .decompress()
+            .ok_or_else(|| CodecError::from("InitialMessage `a` point is invalid"))?;
+
+        let generators: Vec<RistrettoPoint> = generators
+            .into_iter()
+            .map(|g| {
+                CompressedRistretto(g)
+                    .decompress()
+                    .unwrap_or_else(RistrettoPoint::default)
+            })
+            .collect::<Vec<_>>();
+
+        Ok(InitialMessage { a, generators })
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub struct FinalResponse {
-    s: Vec<Scalar>,
+    response: Vec<Scalar>,
 }
 
-#[derive(Clone)]
+impl Encode for FinalResponse {
+    #[inline]
+    fn size_hint(&self) -> usize {
+        self.response.len() * 32
+    }
+
+    fn encode_to<W: Output>(&self, dest: &mut W) {
+        let response: Vec<[u8; 32]> = self
+            .response
+            .clone()
+            .into_iter()
+            .map(|g| *g.as_bytes())
+            .collect::<Vec<_>>();
```

Comment:
Performance: Use references & avoid intermediate copies.
```suggestion
        let response = self.response
            .iter()
            .map(|g| g.as_bytes())
            .collect::<Vec<_>>();
```